Fix chaining of pipelines

lschr opened the pull request:
Chaining of pipelines does not work. The problem is that each processing function "further down the line" gets called multiple times. Some code to illustrate the problem:

``` python
import slicerator

@slicerator.pipeline
def blah(image):
    print("blah")
    return image

@slicerator.pipeline
def blah2(image):
    print("blah2")
    return image

@slicerator.pipeline
def blah3(image):
    print("blah3")
    return image

# p is some image sequence opened using pims.open()
blah3(blah2(blah(p)))[0];
```

yields

```
blah
blah
blah2
blah
blah2
blah3
```

instead of

```
blah
blah2
blah3
```

This pull request fixes the issue (I think).

lschr committed:
Fix chaining of pipelines

nkeim commented:
Great!! This should include a unit test before it's merged.

caspervdw commented:
Thanks for solving this! I can confirm the issue, but I'm sure yet on your solution.

The original idea was that every child `Slicerator` holds a reference directly to the top ancestor.

In #6, we argued that it would be simpler and better to let every child point to its parent, but we did not implement that because it requires a big refactor and before that @danielballan wanted to take a better look at dask.

The solution you pushed does exactly this, which is great, but I am not sure how this complies with the rest of `slicerator`. It's probably OK but we might want to forward some more parameters to the newly created Slicerator. I agree with @nkeim, we need thorough unittests.

Is there a way to update the `_proc_func` field of the existing slicerator? The following example will create infinite recursion. But something like this I would like better for now.

```
elif isinstance(obj, Slicerator):     
    obj._proc_func = lambda x: func(obj._proc_func(x))
    return obj
```

Or @danielballan is it time for the big refactor?

danielballan commented:
I think it is time to look at dask. I spoke with Jim Crist briefly at the PyData conference last week. I'll get an email thread going next week.

caspervdw commented:
Ok! 

But we should fix this chained pipelines issue in the near future. By reading the code and doing some simple tests, I think the proposed fix is OK, so I say we can merge this before the refactor to fix the functionality.

But it still needs proper unittests. @lschr if necessary I can help making the tests.

lschr committed:
Add test for nested pipelines

Make sure that the each chained pipeline function is only called once.

lschr committed:
Partially fix attribute propagation with pipelines

Methods decorated with @index_attr are not working (yet).

caspervdw commented:
Agreed, if you comment out the failing tests of indexed propagation, then I can merge and make an issue about indexed propagation of attributes through pipelines. We have to take a better look at attribute propagation through pipelines anyway: not every attribute will make sense to propagate.

lschr committed:
Disable tests for indexed attr propagation in pipelines

Propagation of indexed attributes currently does not work. A fix is not
obvious.

caspervdw commented:
Thanks again @lschr , merging and I will take a look on the propagation of indexed attrs soon. After that we should do a release for this fix (agreed @danielballan?)

danielballan commented:
Agreed.
